### ISSUE TITLE
Endpoint Routing and dynamic Edm model creation

### ISSUE THREAD

**Issue by toresenneseth** [OData/WebApi#2059]
Hi all,

I'm trying to add an OData service that support dynamic Edm model creation to our web api that use Endpoint routing. Basically, something like this: https://github.com/OData/ODataSamples/tree/master/WebApiCore/DynamicEdmModelCreation

I'm using the Endpoint nightly build, and "almost" got it working, but there seems to be some issue with the output serialization, because the serialization of my payload is terminated.

When I call my odata endpoint at 
http://localhost:44376/odata/DemoFeed/MySource I get the following response:

```
{"@odata.context":"http://localhost:44376/odata/%7BdataSource%7D/$metadata#MySource","value":[
```
As you can see, the serialization is terminated "mid-air".
I tracked down the reason for this to an error happening during the serialization run by the ODataOutputFormatter, but I do not understand why the error occur.

This is the error message I get:
```
A resource of type 'ns.MySource' was found in a resource set that otherwise has entries of type 'ns.MySource'. In OData, all entries in a resource set must have a common base type
```

Does anyone know what this error message mean, considering it appears that the types match (ns.MySource).

Here's the stack trace:
```
   at Microsoft.OData.ResourceSetWithoutExpectedTypeValidator.ValidateResource(IEdmType itemType)
   at Microsoft.OData.ODataWriterCore.ValidateResourceForResourceSet(ODataResourceBase resource, ResourceBaseScope resourceScope)
   at Microsoft.OData.ODataWriterCore.<>c__DisplayClass121_0.<WriteStartResourceImplementation>b__0()
   at Microsoft.OData.ODataWriterCore.InterceptException(Action action)
   at Microsoft.OData.ODataWriterCore.WriteStartResourceImplementation(ODataResource resource)
   at Microsoft.OData.ODataWriterCore.WriteStart(ODataResource resource)
   at Microsoft.AspNet.OData.Formatter.Serialization.ODataResourceSerializer.WriteResource(Object graph, ODataWriter writer, ODataSerializerContext writeContext, IEdmTypeReference expectedType) in C:\Work\Code\Git\Repos\WebApi\src\Microsoft.AspNet.OData.Shared\Formatter\Serialization\ODataResourceSerializer.cs:line 354
   at Microsoft.AspNet.OData.Formatter.Serialization.ODataResourceSerializer.WriteObjectInline(Object graph, IEdmTypeReference expectedType, ODataWriter writer, ODataSerializerContext writeContext) in C:\Work\Code\Git\Repos\WebApi\src\Microsoft.AspNet.OData.Shared\Formatter\Serialization\ODataResourceSerializer.cs:line 77
   at Microsoft.AspNet.OData.Formatter.Serialization.ODataResourceSetSerializer.WriteResourceSet(IEnumerable enumerable, IEdmTypeReference resourceSetType, ODataWriter writer, ODataSerializerContext writeContext) in C:\Work\Code\Git\Repos\WebApi\src\Microsoft.AspNet.OData.Shared\Formatter\Serialization\ODataResourceSetSerializer.cs:line 147
   at Microsoft.AspNet.OData.Formatter.Serialization.ODataResourceSetSerializer.WriteObjectInline(Object graph, IEdmTypeReference expectedType, ODataWriter writer, ODataSerializerContext writeContext) in C:\Work\Code\Git\Repos\WebApi\src\Microsoft.AspNet.OData.Shared\Formatter\Serialization\ODataResourceSetSerializer.cs:line 87
   at Microsoft.AspNet.OData.Formatter.Serialization.ODataResourceSetSerializer.WriteObject(Object graph, Type type, ODataMessageWriter messageWriter, ODataSerializerContext writeContext) in C:\Work\Code\Git\Repos\WebApi\src\Microsoft.AspNet.OData.Shared\Formatter\Serialization\ODataResourceSetSerializer.cs:line 56
   at Microsoft.AspNet.OData.Formatter.ODataOutputFormatterHelper.WriteToStream(Type type, Object value, IEdmModel model, ODataVersion version, Uri baseAddress, MediaTypeHeaderValue contentType, IWebApiUrlHelper internaUrlHelper, IWebApiRequestMessage internalRequest, IWebApiHeaders internalRequestHeaders, Func`2 getODataMessageWrapper, Func`2 getEdmTypeSerializer, Func`2 getODataPayloadSerializer, Func`1 getODataSerializerContext) in C:\Work\Code\Git\Repos\WebApi\src\Microsoft.AspNet.OData.Shared\Formatter\ODataOutputFormatterHelper.cs:line 208
   at Microsoft.AspNet.OData.Formatter.ODataOutputFormatter.WriteResponseBodyAsync(OutputFormatterWriteContext context, Encoding selectedEncoding) in C:\Work\Code\Git\Repos\WebApi\src\Microsoft.AspNetCore.OData\Formatter\ODataOutputFormatter.cs:line 228
```

Is there a bug in the ResourceSetWithoutExpectedTypeValidator?
Is this caused by trying to use Endpoint routing and dynamic model creation?

Basically, I'm looking for some hint to point me on the right path to track this down.

-Thanks!

**Comment by xuzhg** [comment]
@toresenneseth Can you share your changes in a repo?

**Comment by toresenneseth** [comment]
Yes,

I have a demo based on your AspNetCore3xEndpointSample here:
https://github.com/toresenneseth/AspNetCoreODataEndpointRouting

I might be going about this totally wrong, but it seems to be so close to working if it wasn't for the validation error that's confusing me.

-Thanks!

**Comment by xuzhg** [comment]
@toresenneseth 

The problem is that you create the model multiple time for one request in `DataSourceProvider.cs`

I have a quick change as follows about the "GetEdmModel()" method in "DataSourceProvider.cs"

```C#
        public static IEdmModel GetEdmModel(HttpRequest request)
        {
            if (request.HttpContext.Items.TryGetValue("__ODataEdmModel", out object edmModel))
            {
                return (IEdmModel)edmModel;
            }

            EdmModel model = new EdmModel();
            EdmEntityContainer container = new EdmEntityContainer("ns", "container");
            model.AddElement(container);

            var dataSource = GetDataSource(request);
            dataSource.GetModel(model, container);

            request.HttpContext.Items.Add("__ODataEdmModel", model);
            return model;
        }

```

It works as:

![image](https://user-images.githubusercontent.com/9426627/74792983-d4047500-5273-11ea-827c-5262ba199fd3.png)

By the way, it seems you are using my old "Endpoint routing" codes. Would you please help to update the "WebApi" folder using my latest codes at: https://github.com/xuzhg/WebApi/tree/EnableEndpointRouting

And do more testing? Thanks a lot.

**Comment by toresenneseth** [comment]
This is beautiful, thanks! 
I just pulled down the latest version of the code and updated the repository, and things continue to work.

I noticed that the Edm model was created several times, but didn't realize that's what caused the validation to fail. I tried with both Transient and Scoped lifetimes, but there are probably several scopes created during a request. So I guess there must be some kind of reference equality going on in the validation? 

Anyway, I'll go ahead and implement these changes in our main project and close this issue if everything works fine.

I'll leave the repository available as a reference for others.

Btw, really appreciate your efforts adding support for Endpoint Routing before .NET 5! Great job!

-Thanks

**Comment by toresenneseth** [comment]
I can confirm that I got things working in production with endpoint routing and dynamic model creation. I'm currently using the nightly Microsoft.AspNetCore.OData 7.4.0-Nightly202001292116

https://www.myget.org/feed/webapinetcore/package/nuget/Microsoft.AspNetCore.OData/7.4.0-Nightly202001292116

I'll leave it to use to close the issue and apply whatever labeling you see fit.

- Thanks!

**Comment by xuzhg** [comment]
Thanks and close this. FYI. 7.4.beta Nuget is just released at https://www.nuget.org/packages/Microsoft.AspNetCore.OData/7.4.0-beta

Please try it and let us know any new issues/concerns. Thanks

**Comment by huzaifak** [comment]
> This is beautiful, thanks!
> I just pulled down the latest version of the code and updated the repository, and things continue to work.
> 
> I noticed that the Edm model was created several times, but didn't realize that's what caused the validation to fail. I tried with both Transient and Scoped lifetimes, but there are probably several scopes created during a request. So I guess there must be some kind of reference equality checks going on in the validation?
> 
> Anyway, I'll go ahead and implement these changes in our main project and close this issue if everything works fine.
> 
> I'll leave the repository available as a reference for others.
> 
> Btw, really appreciate your efforts adding support for Endpoint Routing before .NET 5! Great job!
> 
> -Thanks

@xuzhg , @toresenneseth 
Hello, The sample repo uses the MVC routing, were you able to do the same custom routing using the endpoint routing. can you share any sample for the same?

**Comment by toresenneseth** [comment]
Hi,

I used endpoint routing in my sample repo here:

https://github.com/toresenneseth/AspNetCoreODataEndpointRouting/blob/master/AspNetCore3xEndpointSample/Startup.cs

**Comment by huzaifak** [comment]
@toresenneseth thanks for the sample will check it out.

**Comment by huzaifak** [comment]
@toresenneseth i looked at the sample, your original DynamicEdmModelCreation was using Custom Odata  path with MVC, the sample you shared doesn't uses custom path. i have not been able to find anyways to register custom odata route with endpoint routing. Is this something supported? any clue will be helpful. Thanks.

**Comment by cvraman** [comment]
@toresenneseth : I was able to get your sample working. Only thing is, EnableQuery is not working with dynamic edm model.

@xuzhg : Is this scenario supported using OData ?
